caches/infinispan/src/main/java/org/commonjava/maven/galley/cache/infinispan/FastLocalCacheProvider.java

```diff
@@ -316,71 +316,113 @@ public boolean isFile( ConcreteResource resource )
+        {
+            @Override
+            public void run()
+            {
+                InputStream nfsIn = null;
+                OutputStream localOut = null;
+
+                try
+                {
+                    cacheTxMgr.begin();
+                    nfsOwnerCache.getAdvancedCache().lock( pathKey );
+                    File nfsFile = getNFSDetachedFile( resource );
+                    if ( !nfsFile.exists() )
+                    {
+                        logger.debug( "NFS cache does not exist too." );
+                        return;
+                    }
+                    nfsIn = new FileInputStream( nfsFile );
+                    localOut = plCacheProvider.openOutputStream( resource );
+                    IOUtils.copy( nfsIn, localOut );
+                    logger.debug( "NFS copy to local cache done." );
+                }
+                catch ( NotSupportedException | SystemException | IOException e )
+                {
+                    if ( e instanceof IOException )
+                    {
+                        final String errorMsg =
+                                String.format( "[galley] got i/o error when doing the NFS->Local copy for resource %s",
+                                               resource.toString() );
+                        logger.warn( errorMsg, e );
+                    }
+                    else
+                    {
+                        final String errorMsg = String.format(
+                                "[galley] Cache TransactionManager got error, locking key is %s, resource is %s",
+                                pathKey, resource.toString() );
+                        logger.error( errorMsg, e );
+                        throw new IllegalStateException( errorMsg, e );
+                    }
+                }
+                finally
+                {
+                    try
+                    {
+                        cacheTxMgr.rollback();
+                    }
+                    catch ( SystemException e )
+                    {
+                        final String errorMsg = String.format(
+                                "[galley] Cache TransactionManager rollback got error, locking key is %s", pathKey );
+                        logger.error( errorMsg, e );
+                    }
+                    IOUtils.closeQuietly( nfsIn );
+                    IOUtils.closeQuietly( localOut );
+                    synchronized ( copyLock )
+                    {
+                        canStreamOpen.set( true );
+                        copyLock.notifyAll();
+                    }
+                }
+            }
+        };
+        // This lock is used to control the concurrent operations on the resource, like concurrent delete and read/write.
+        // Use "this" as lock is heavy, should think about use the transfer for the resource as the lock for each thread
         synchronized ( getTransfer( resource ) )
```

Comment:
I wonder if we need the whole thing (starting the copy thread and waiting to open an InputStream for the user) inside this synchronized block. It seems like synchronizing on the Transfer instance like that would avoid the case where two users race to initiate the copy from NFS into the local cache, and wind up stepping on one another (or blocking on competing writes)...